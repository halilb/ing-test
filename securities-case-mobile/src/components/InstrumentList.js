import React, { PureComponent } from 'react';
import { FlatList, View } from 'react-native';
import { PropTypes } from 'prop-types';

import InstrumentRow from './InstrumentRow';
import RowSeparator from './RowSeparator';

export default class InstrumentList extends PureComponent {
  static propTypes = {
    instruments: PropTypes.array,
    style: View.propTypes.style,
  };

  keyExtractor = item => item.code;

  renderItem = ({ item }) => {
    const { code, roundedPrice, priceDiffPercentage } = item;
    return (
      <InstrumentRow
        key={`instrument_${code}`}
        id={code}
        code={code}
        price={roundedPrice}
        priceDiffPercentage={priceDiffPercentage}
      />
    );
  };

  render() {
    const { style, instruments } = this.props;

    return (
      <FlatList
        style={style}
        keyExtractor={this.keyExtractor}
        data={instruments}
        renderItem={this.renderItem}
        ItemSeparatorComponent={RowSeparator}
      />
    );
  }
}
