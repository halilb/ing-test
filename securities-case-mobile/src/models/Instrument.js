import { computed } from 'mobx';

export default class Instrument {
  /**
   * code is the unique identifier for the instrument.
   */
  code = null;

  /*
   * instrument's initial price is used to calculate
   * difference with it's current price.
   * it is immutable.
   */
  initialPrice = null;

  /*
   * current price of the instrument.
   */
  price = null;

  /*
   * price update time,
   * provided by the API.
   * not used in the UI currently.
   */
  lastPriceTime = null;

  constructor(code, price, lastPriceTime) {
    this.code = code;
    this.initialPrice = price;
    this.price = price;
    this.lastPriceTime = lastPriceTime;
  }

  // displayed price
  @computed get roundedPrice() {
    return Math.round(this.price * 100) / 100;
  }

  @computed get priceDiffPercentage() {
    const percentage = this.price / this.initialPrice * 100 - 100;
    return Math.round(percentage * 100) / 100;
  }
}
