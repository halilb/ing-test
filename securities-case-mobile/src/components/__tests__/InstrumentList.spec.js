import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import Instrument from '../../models/Instrument';
import InstrumentList from '../InstrumentList';

it('renders an empty list correctly', () => {
  const snapshot = renderer
    .create(
      <InstrumentList style={{ backgroundColor: 'red' }} instruments={[]} />
    )
    .toJSON();
  expect(snapshot).toMatchSnapshot();
});

it('renders instruments correctly', () => {
  const IngInsturement = new Instrument('ING', 14.523253412, 1497888536011);
  const AppleInstrument = new Instrument('AAPL', 258.3693474, 1497888703132);

  const snapshot = renderer
    .create(<InstrumentList instruments={[IngInsturement, AppleInstrument]} />)
    .toJSON();
  expect(snapshot).toMatchSnapshot();
});
