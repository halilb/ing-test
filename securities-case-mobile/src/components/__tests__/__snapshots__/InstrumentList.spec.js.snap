// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an empty list correctly 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onViewableItemsChanged={undefined}
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "backgroundColor": "red",
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;

exports[`renders instruments correctly 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={
    Array [
      Instrument {
        "code": "ING",
        "initialPrice": 14.523253412,
        "lastPriceTime": 1497888536011,
        "price": 14.523253412,
      },
      Instrument {
        "code": "AAPL",
        "initialPrice": 258.3693474,
        "lastPriceTime": 1497888703132,
        "price": 258.3693474,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onViewableItemsChanged={undefined}
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={undefined}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingHorizontal": 0,
            "paddingVertical": 8,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "fontWeight": "bold",
            }
          }
        >
          ING
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 0,
              "flexDirection": "row",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#B8B8B8",
                "marginRight": 8,
              }
            }
          >
            14.52
          </Text>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 2,
                  "paddingVertical": 2,
                  "width": 52,
                },
                Object {
                  "backgroundColor": "#55EDB5",
                },
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              disabled={false}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "white",
                  "fontSize": 12,
                }
              }
            >
              0
              %
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "rgb(200, 199, 204)",
            "height": 0.5,
            "marginLeft": 8,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingHorizontal": 0,
            "paddingVertical": 8,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "fontWeight": "bold",
            }
          }
        >
          AAPL
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 0,
              "flexDirection": "row",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#B8B8B8",
                "marginRight": 8,
              }
            }
          >
            258.37
          </Text>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 2,
                  "paddingVertical": 2,
                  "width": 52,
                },
                Object {
                  "backgroundColor": "#55EDB5",
                },
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              disabled={false}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "white",
                  "fontSize": 12,
                }
              }
            >
              0
              %
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
